homeassistant/components/mqtt/__init__.py

```diff
@@ -644,6 +644,8 @@ def _match_topic(subscription, topic):
     if subscription.endswith('#'):
         subscription = subscription[:-2]
         suffix = "(.*)"
+    if subscription.startswith('$'):
```

Comment:
I don't think that this is the correct solution. Instead, we should call [`re.escape()`](https://docs.python.org/3/library/re.html#re.escape) on each part of the topic that's not `+`.